File: Daily/3887-minimum-cost-path-with-edge-reversals/minimum-cost-path-with-edge-reversals.cpp
class Solution {
public:
    int minCost(int n, vector<vector<int>>& edges) {
        vector<vector<pair<int,int>>> g(n);
        vector<vector<pair<int,int>>> in(n);

        for (auto &e : edges) {
            int u = e[0], v = e[1], w = e[2];
            g[u].emplace_back(v, w); 
            in[v].emplace_back(u, w);
        }

        using T = pair<int,int>; // cost, node
        priority_queue<T, vector<T>, greater<T>> pq;

        vector<int> dist(n, INT_MAX);
        dist[0] = 0;
        pq.emplace(0, 0);
        while (!pq.empty()) {
            auto [cost, u] = pq.top();
            pq.pop();

            if (u == n - 1) {
                return cost;
            }
            if (cost > dist[u]) continue;

            for (auto [v, w] : g[u]) {
                int nxtCost = cost + w;
                if (nxtCost < dist[v]) {
                    dist[v] = nxtCost;
                    pq.emplace(nxtCost, v);
                }
            }

            
            for (auto [v, w] : in[u]) {
                int nxtCost = cost + 2 * w;
                if (nxtCost < dist[v]) {
                    dist[v] = nxtCost;
                    pq.emplace(nxtCost, v);
                }
            }
        }

        return -1;
    }
};

File: Daily/3887-minimum-cost-path-with-edge-reversals/minimum-cost-path-with-edge-reversals.py
class Solution:
    def minCost(self, n: int, edges: List[List[int]]) -> int:
        g = [[] for _ in range(n)]
        for x, y, w in edges:
            g[x].append((y, w))
            g[y].append((x, 2*w))
        dist = [inf] * n
        dist[0] = 0
        h = [(0,0)]
        while h:
            dis, node = heappop(h)
            if dis > dist[node]:
                continue
            if node == n - 1:
                return dis
            for y, wt in g[node]:
                if wt + dis < dist[y]:
                    dist[y] = wt + dis
                    heappush(h, (wt + dis, y))
        return -1
            